### PR TITLE
Feature/jolokia proxy mode

### PR DESCRIPTION
**Pull request by saiello**
Added jolokia proxy mode.

### CODE REVIEW

File: plugins/inputs/jolokia/jolokia.go [hunk 1]
```diff
@@ -41,20 +44,32 @@ func (c JolokiaClientImpl) MakeRequest(req *http.Request) (*http.Response, error
...
 func (j *Jolokia) SampleConfig() string {
 	return `
-  ## This is the context root used to compose the jolokia url
-  context = "/jolokia/read"
+  # This is the context root used to compose the jolokia url
+  context = "/jolokia"
 
-  ## List of servers exposing jolokia read service
+  # This specifies the mode used
+  # mode = "proxy"
```
Comment (sparrc) [inline plugins/inputs/jolokia/jolokia.go]: What are the options besides "proxy"?

File: plugins/inputs/jolokia/jolokia.go [hunk 2]
```diff
@@ -41,20 +44,32 @@ func (c JolokiaClientImpl) MakeRequest(req *http.Request) (*http.Response, error
...
+	Mode    string
 	Servers []Server
 	Metrics []Metric
+	Proxy   Server
 }
 
 func (j *Jolokia) SampleConfig() string {
 	return `
-  ## This is the context root used to compose the jolokia url
-  context = "/jolokia/read"
```
Comment (sparrc) [inline plugins/inputs/jolokia/jolokia.go]: why are you changing the default context?

Comment (saiello) [inline plugins/inputs/jolokia/jolokia.go]: '/jolokia' is actually the context. 'read' is the operation. I started to use POST request to support proxy mode and I moved the operation type into the request body.

File: plugins/inputs/jolokia/jolokia.go (hunk 1 above)
Comment (saiello) [inline plugins/inputs/jolokia/jolokia.go]: Currently any value different to "proxy", make the input plugin executing in agent mode.

File: plugins/inputs/jolokia/jolokia.go [hunk 3]
```diff
@@ -64,30 +79,28 @@ func (j *Jolokia) SampleConfig() string {
...
+    mbean  = "java.lang:type=Memory"
+    attribute = "HeapMemoryUsage"
+
   ## This collect thread counts metrics.
   [[inputs.jolokia.metrics]]
     name = "thread_count"
-    jmx  = "/java.lang:type=Threading/TotalStartedThreadCount,ThreadCount,DaemonThreadCount,PeakThreadCount"
- 
+    mbean  = "java.lang:type=Threading"
+		attribute = "TotalStartedThreadCount,ThreadCount,DaemonThreadCount,PeakThreadCount"
```
Comment (sparrc) [inline plugins/inputs/jolokia/jolokia.go]: this one (and one below) should be indented by 2 spaces, not tabs

**Comment by sparrc** [comment]
This looks like a good change to the jolokia plugin, but it is a breaking change as well.

Will the POST requests to jolokia work correctly for all users out of the box?

File: plugins/inputs/jolokia/jolokia.go (hunk 3 above)
Comment (saiello) [inline plugins/inputs/jolokia/jolokia.go]: I fix it

**Comment by saiello** [comment]
POST requests are more powerful and have less escaping issue than GET ones.
https://jolokia.org/reference/html/protocol.html#request-response 

Yes it is a breaking change in configuration, but it is trivial to change the old 'jmx' with new Mbean/Attribute/Path parts.

Would you make it working even after update keeping the old configuration?

**Comment by sparrc** [comment]
No, I think it's OK if we make it a breaking change. I just wanted to confirm

File: plugins/inputs/jolokia/jolokia.go [hunk 4]
```diff
@@ -132,37 +226,37 @@ func (j *Jolokia) Gather(acc telegraf.Accumulator) error {
 		tags["port"] = server.Port
```
Comment (sparrc) [inline plugins/inputs/jolokia/jolokia.go]: @saiello I should have caught this in the original plugin, but here you are overwriting the telegraf-wide `host` tag, could you change these tags:

- server
- port
- host

to:

- server_name
- server_port
- server_host

**Comment by sparrc** [comment]
@saiello you need to run `go fmt ./...`

**Comment by sparrc** [comment]
thanks @saiello, please rebase your changes and update CHANGELOG.md with this change, as well as writing up something about it being a breaking change under "Release Notes", and then I can merge.

Please also ping me when you're done with this, as I don't get notifications on file changes

**Comment by sparrc** [comment]
this change will now also close #1050

**Comment by saiello** [comment]
@sparrc I edited the release note. Would you require some kind of "migration guide" in jolokia README?

**Comment by darsh221** [comment]
@sparrc When this will be merged to master?

**Comment by sparrc** [comment]
by the end of the week

**Comment by darsh221** [comment]
Cool. Thank you for the update.

**Pull request closed without merging**